include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+#ifndef __NBL_I_SYSTEM_H_INCLUDED__
+#define __NBL_I_SYSTEM_H_INCLUDED__
+
+#include <variant>
+#include "nbl/core/IReferenceCounted.h"
+#include "nbl/system/ISkippableAsyncQueueDispatcher.h"
+#include "nbl/system/IFileArchive.h"
+#include "nbl/system/IWriteFile.h"
+
+namespace nbl {
+namespace system
+{
+
+class ISystem final : public core::IReferenceCounted
+{
+public:
+    class ISystemCaller : core::IReferenceCounted
+    {
+    protected:
+        virtual ~ISystemCaller() = default;
+
+    public:
+        virtual core::smart_refctd_ptr<IReadFile> createReadFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual core::smart_refctd_ptr<IWriteFile> createWriteFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual uint32_t read(IReadFile* file, void* buffer, size_t offset, size_t size) = 0;
+        virtual uint32_t write(IWriteFile* file, const void* buffer, size_t offset, size_t size) = 0;
+    };
+
+protected:
+    struct SRequestType;
+    class CAsyncQueue;
+
+    class SFutureBase
+    {
+        friend ISystem;
+
+    protected:
+        bool valid_flag = false;
+        SRequestType* request;
+        CAsyncQueue* asyncQ;
+
+    public:
+        // SFuture is non-copyable and non-movable
+        SFutureBase(const SFutureBase&) = delete;
+
+        ~SFutureBase()
+        {
+            request->set_skip();
+        }
+
+        bool ready() const { return !request || request->ready; }
+        bool valid() const { return valid_flag; }
+
+        void wait()
+        {
+            asyncQ->waitForRequestCompletion(request[0]);
+        }
+    };
+
+    template <typename T>
+    class SFuture : public SFutureBase
+    {
+        alignas(T) uint8_t storage[sizeof(T)];
+
+    public:
+        ~SFuture()
+        {
+            if (valid_flag)
+                getStorage()->~T();
```

Comment:
if you're going to provide a facility to call the placement destructor, then provide a facility to have the placement constructor called, maybe roll it up into a `notify` method
```
template<class... Args>
void notify(Args&&... args)
{
    new (getStorage()) T(std::forward<Args>(args)...);
    SFutureBase::notify(true);
}

void fail()
{
    SFutureBase::notify(false);
}
```

Reply:
Hmm actually maybe coupling the notification with object creation is not good.

Because it would be good to create the object at the start of the processing. So then it would make sense to have
```cpp
template<T>
struct SFutureStorage
{
protected:
    alignas(T) uint8_t storage[sizeof(T)];
public:
    template<class... Args>
    inline T& construct(Args&&... args)
    {
        return new (storage) T(std::forward<Args>(args)...);
    }
};

template<T>
struct SFuture : SFutureStorage<T>, SFutureBase
{
    inline void notify(bool success)
    {
        if (!success)
            getStorage()->~T();
        valid_flag = success;
        request = nullptr;
    }
};

```

Reply:
resolved.